src/weights.rs

```diff
@@ -0,0 +1,241 @@
+//! Code for working with distance relationships in taxonomic trees
+//! and tracking weights assigned to different nodes in the tree.
+use std::collections::{BinaryHeap, HashMap, HashSet};
+use std::fmt::{Debug, Display};
+use std::hash::{BuildHasher, Hash};
+use std::iter::Sum;
+
+use crate::taxonomy::Taxonomy;
+use crate::Result;
+
+/// Calculates the summed weight of every path to root for a set of node
+/// weights given a corresponding taxonomy.
+///
+/// This is like `maximum_weighted_path`, but rather than just returning the
+/// _best_ path, it returns all of the paths.
+pub fn all_weighted_paths<'t, D: 't, T: 't, W: 't, S: BuildHasher>(
+    taxonomy: &'t impl Taxonomy<'t, T, D>,
+    weights: &HashMap<T, W, S>,
+) -> Result<Vec<(T, W)>>
+where
+    D: Debug + PartialOrd + Sum,
+    T: Clone + Debug + Display + Hash + PartialEq + Ord,
+    W: Clone + Debug + Ord + PartialOrd + PartialEq + Sum,
```

Comment:
Looking forward to https://doc.rust-lang.org/unstable-book/language-features/trait-alias.html